File: Py-Solutions/Problem-4.py
# Find the largest palindrome made from the product of two 3-digit numbers.

def isPalindrome(num):
  str_num = str(num)
  rev_num = str_num[-1]
  for i in range(len(str_num)-2, -1, -1):
    rev_num += str_num[i]
  if rev_num == str_num:
    return True
  return False

def largestPalindrome(limit):
  max = [0,0,0]
  for i in range(900,limit):
    for j in range(900,limit):
      if isPalindrome(i * j) and i * j > max[0]:
        max = [i*j,i,j]
  return(max)

print(largestPalindrome(1000)[0])

File: R-Solutions/Problem-4.R
# Find the largest palindrome made from the product of two 3-digit numbers.

isPalindrome <- function(num) {
  pal_arr <- c()
  x = 1
  while(num > 9) {
    pal_arr <- append(pal_arr, (num %% 10**x) / 10**(x-1))
    num <- num - num %% 10**x
    x <- x + 1
  }
  return(sum(pal_arr == rev(pal_arr)) == length(pal_arr))
}

largestPalindrome <- function(limit) {
  max_palindrome <- 0
  for (i in ((limit/2):limit)) {
    for (j in ((limit/2):limit)) {
      if (isPalindrome(i*j) && i*j > max_palindrome) {
        max_palindrome <- i*j
      }
    }
  }
  return(max_palindrome)
}

largestPalindrome(1000)
